File: Excel/598 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/598 Palindromic Step Harshad Numbers.xlsx"
test  = read_excel(path, range = "A1:A10")

is_palindromic = function(number) {
  as.character(number) == str_c(rev(str_split(as.character(number), "")[[1]]), collapse = "")
}

is_step = function(number) {
  digits = as.integer(str_split(as.character(number), "")[[1]])
  all(abs(diff(digits)) == 1)
}

is_harshad = function(number) {
  digits = as.integer(str_split(as.character(number), "")[[1]])
  number %% sum(digits) == 0
}

result = 10:10000000 %>% 
  keep(is_palindromic) %>% 
  keep(is_step) %>% 
  keep(is_harshad)

all.equal(result, test$`Answer Expected`)
# [1] TRUE

File: Excel/500-599/598/598 Challenge.py
import pandas as pd

path = "598 Palindromic Step Harshad Numbers.xlsx"
test = pd.read_excel(path, usecols="A", nrows=10).squeeze().tolist()

def is_palindromic(num):
    s = str(num)
    return s == s[::-1]

def is_step(num):
    digits = list(map(int, str(num)))
    return all(abs(digits[i] - digits[i+1]) == 1 for i in range(len(digits) - 1))

def is_harshad(num):
    digit_sum = sum(map(int, str(num)))
    return num % digit_sum == 0

range_limit = 10000000
range_start = 10

palindromic = [n for n in range(range_start, range_limit + 1) if is_palindromic(n)]
step_palindromic = [n for n in palindromic if is_step(n)]
result = [n for n in step_palindromic if is_harshad(n)]

print(result == test) # True
